### PR TITLE
finished tasks 01-03

### PR DESCRIPTION
**Pull request by Pavel-Lapatin**
Artsiom, Hi. Send you finished 1-3 tasks of lab04 SQL.

### CODE REVIEW

**Comment by a-yandulski** [comment]
Hi @Pavel-Lapatin , please setup file encoding to UTF-8 in SSMS (File > Advanced Save Options) if you use it or add .gitattributes file with line:
`
.sql diff
`

**Comment by a-yandulski** [comment]
1) It is considered a bad practice to use **\*** in the SELECT. Columns may change in future and your code will stop working 
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task02.sql#L8
2) Although this works, a better (right) solution exists: [DATEDIFF](https://docs.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql)
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task02.sql#L17
3) Unexpected column in the result - **Gender**
https://github.com/Godel-Mastery-v2/p-lapatin/blob/2c3bf5f1f94de3ecf9b4480be692554666e9aea5/lab04/task02.sql#L24
4) Please fix coding style
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task02.sql#L35
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task02.sql#L45
5) It's possible to simplify the query to use one-line statement for 'Last Word'
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task02.sql#L59
6) It's better to start the **JOIN** statement on a new line.
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L9
7) I'd go with **[]** brackets for column alias as single quotes here look like a string and harm readability  
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L8
5) DepartmentID column is missing
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L45
9) Unnecessary join here. Any idea why (we can discuss this on Skype)?
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L47
10) Unexpected column in the result - **RateChangeDate**
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L56
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L66
11) Curved brackets are not needed here
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L60
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L85
12) So, what the best query?  Lets discuss this on Skype
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L54
13) Please rename columns according to the task description
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L57
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L59
14)  Last two columns don't have names
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L104
15)  What can be used instead of CONVERT function?
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L104
16) Wrong name of the 3rd column. Please check expected output in the lab
https://github.com/Godel-Mastery-v2/p-lapatin/blob/1ca09499188dcff79d5717f867aa43ed9503e96d/lab04/task03.sql#L110

File: lab04/task03.sql [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+       INNER JOIN HumanResources.EmployeePayHistory AS eph 
+	   ON e.BusinessEntityID = eph.BusinessEntityID
+       INNER JOIN HumanResources.EmployeeDepartmentHistory AS edh 
+	   ON e.BusinessEntityID = edh.BusinessEntityID
+ WHERE edh.EndDate IS NULL
+ GROUP BY e.BusinessEntityID, e.JobTitle
+HAVING COUNT(eph.Rate) > 1
+GO
+
+	--Исправленный запрос 3. С учетом "Если сотрудник больше не работает в отделе — не учитывать такие данные."
```
Comment (a-yandulski) [inline lab04/task03.sql]: Hi @Pavel-Lapatin this case is not needed. Discussed with @RamanBut-Husaim, he will update the task description.

File: lab04/task03.sql [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	   ON e.BusinessEntityID = eph.BusinessEntityID
+set statistics time off;
+
+
+--6. Вывести на экран почасовые ставки сотрудников, с указанием максимальной ставки для каждого отдела в столбце MaxInDepartment. 
+--В рамках каждого отдела разбейте все ставки на группы таким образом, чтобы ставки с одинаковыми значениями входили в состав одной группы. 
+
+SELECT d.Name, 
+       e.BusinessEntityID,
+	   MAX(eph.Rate) AS maxRate,
```
Comment (a-yandulski) [inline lab04/task03.sql]: @Pavel-Lapatin , 2 issues here:
1) **MAX(eph.Rate)** doesn't work here. Can you tell me why?
2) **maxRate** in not a valid column name

Comment (Pavel-Lapatin) [reply lab04/task03.sql]: Because we group select by Rate and in this case Rate will be a distinct value. So it doesn't make any sense to find MAX.

Comment (Pavel-Lapatin) [reply lab04/task03.sql]: The only case the Group by clause will have more than one tupple, if one employee worked in different departments with the same name and has the same Rate there, but even in that case Rates will be  the same. Am i right?

Comment (Pavel-Lapatin) [reply lab04/task03.sql]: And department.name cannot be the same as it has unique index.

File: lab04/task5.0.sql [hunk 3]
```diff
@@ -0,0 +1,132 @@
+﻿--Task 5.
+
+
+--1. Создайте таблицу dbo.Person_New с такой же структурой как dbo.Person.
+
+USE [AdventureWorks2017]
+GO
+
+/****** Object:  Table [dbo].[Person_New]    Script Date: 12/4/2017 4:02:24 PM ******/
+SET ANSI_NULLS ON
```
Comment (a-yandulski) [inline lab04/task5.0.sql]: @Pavel-Lapatin, looks like you've used **"Script Table as"** here. Please re-write CREATE TABLE statement without unnecessary commands\options.  The common rule is "Never add options\commands that you don't need, don't understand or  are not requested to add".

File: lab04/task4.1.sql [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+ ALTER TABLE dbo.Person
+ ADD CONSTRAINT DF_Person_Title DEFAULT N'N/A' FOR Title
+ GO
+
+ --5. Заполните новую таблицу данными из Person.Person только для тех сотрудников, 
+ --которые существуют в таблице HumanResources.Employee, исключив сотрудников из отдела Finance.
+
+INSERT INTO dbo.Person (BusinessEntityID, PersonType, NameStyle, FirstName, MiddleName, LastName, Suffix, EmailPromotion, ModifiedDate)
+SELECT p.BusinessEntityID, p.PersonType, p.NameStyle, p.FirstName, p.MiddleName, p.LastName, p.Suffix, p.EmailPromotion, p.ModifiedDate
```
Comment (a-yandulski) [inline lab04/task4.1.sql]: @Pavel-Lapatin, **p.Title** should also be inserted

File: lab04/task4.2.sql [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ ALTER TABLE dbo.StateProvince
+ ADD CONSTRAINT DF_StateProvince_ModifiedDate DEFAULT GETDATE() FOR ModifiedDate
+ GO
+
+ --5. Заполните новую таблицу данными из Person.StateProvince. 
+ --Выберите для вставки только те данные, где имя штата/государства совпадает с именем страны/региона в таблице CountryRegion.
+
+ GO
+ INSERT INTO dbo.StateProvince (StateProvinceID, StateProvinceCode, CountryRegionCode, IsOnlyStateProvinceFlag, Name, TerritoryID)
+ SELECT  sp.StateProvinceID, sp.StateProvinceCode, sp.CountryRegionCode, sp.IsOnlyStateProvinceFlag, sp.Name, sp.TerritoryID
```
Comment (a-yandulski) [inline lab04/task4.2.sql]: @Pavel-Lapatin , **ModifiedDate** should also be inserted

File: lab04/task5.0.sql [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+ALTER TABLE [dbo].[Person_New]  WITH CHECK ADD  CONSTRAINT [CHK_Person_New_MiddelName] CHECK  (([MiddleName]='L' OR [MiddleName]='J'))
+GO
+
+ALTER TABLE [dbo].[Person_New] CHECK CONSTRAINT [CHK_Person_New_MiddelName]
+GO
+
+--2. Добавьте в таблицу dbo.Person поле Salutaion типа nvarchar длинной в 80 символов.
+
+ALTER TABLE dbo.Person_New
```
Comment (a-yandulski) [inline lab04/task5.0.sql]: @Pavel-Lapatin, the description states that table should be **dbo.Person** not dbo.Person_New. If you're in doubt please raise a question in Skype\Discord.

File: lab04/task5.0.sql [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	                                               WHEN  e.Gender = 'M' THEN 'Mr.'
+												   ELSE 'Ms.'
+	                                               END
+  FROM dbo.Person AS p 
+	   INNER JOIN HumanResources.Employee as e 
+	   ON p.BusinessEntityID = e.BusinessEntityID
+GO
+ --4. Обновине поле Salutation в таблице dbo.Person значением, которое равно объединению полей Title и FirstName из той же таблицы. 
+ --В качестве разделителя использовать пробел. Например, для Dana Burnell Salutation будет выглядеть как Ms. Dana.
+UPDATE dbo.Person_New
```
Comment (a-yandulski) [inline lab04/task5.0.sql]: @Pavel-Lapatin, Should be **dbo.Person**

File: lab04/task5.0.sql [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+   SET Salutation = CONCAT(Title, ' ', FirstName) 
+GO
+
+--5. Удалите данные из dbo.Person, где значение поля Salutation превысило 10 символов.
+DELETE FROM dbo.Person_New
```
Comment (a-yandulski) [inline lab04/task5.0.sql]: @Pavel-Lapatin, Should be **dbo.Person**

File: lab04/task5.0.sql [hunk 9]
```diff
@@ -60,38 +50,48 @@ INSERT INTO dbo.Person_New (BusinessEntityID,
...
 	   INNER JOIN HumanResources.Employee as e 
 	   ON p.BusinessEntityID = e.BusinessEntityID
 GO
- --4. Обновине поле Salutation в таблице dbo.Person значением, которое равно объединению полей Title и FirstName из той же таблицы. 
+ --4. Обновите поле Salutation в таблице dbo.Person значением, которое равно объединению полей Title и FirstName из той же таблицы. 
  --В качестве разделителя использовать пробел. Например, для Dana Burnell Salutation будет выглядеть как Ms. Dana.
-UPDATE dbo.Person_New
-   SET Salutation = CONCAT(Title, ' ', FirstName) 
+UPDATE p1
+   SET p1.Title = p2.Title
```
Comment (a-yandulski) [inline lab04/task5.0.sql]: @Pavel-Lapatin, mm... don't think this is right. The description states that Title should be from the same table (dbo.Person). If you are worried about NULL try remember hint with NULL + string in SQL, or use [LTRIM](https://docs.microsoft.com/en-us/sql/t-sql/functions/ltrim-transact-sql) trim leading spaces when title is NULL

**Comment by pryabov** [comment]
Good day guys. If this PR is ok and tasks are completed can this PR be merged to `master`?

**Pull request merged**